worker.go

```diff
@@ -2,155 +2,188 @@ package main
+				log.Printf("failed to find a certificate for %v: %v", ingress.CertHostname(), err)
 				continue
 			}
-			log.Printf("using autopilot mode for ingress %v", ingress)
-			uniqueARNs[arn] = append(uniqueARNs[arn], ingress)
+		}
+		if item, ok := model[certificateARN]; ok {
+			item.ingress = ingress
+		} else {
+			model[certificateARN] = &managedItem{ingress: ingress}
 		}
 	}
-	return uniqueARNs
+	return model
 }
 
-func findCertARNForIngress(awsAdapter *aws.Adapter, ingress *kubernetes.Ingress, serverCerts []*aws.CertDetail) (string, error) {
-	acmCert, err := awsAdapter.FindBestMatchingCertificate(serverCerts, ingress.CertHostname())
+func discoverCertificateAndUpdateIngress(certsProvider certs.CertificatesProvider, ingress *kubernetes.Ingress) (string, error) {
+	knownCertificates, err := certsProvider.GetCertificates()
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("discoverCertificateAndUpdateIngress failed to obtain certificates: %v", err)
 	}
-	ingress.SetCertificateARN(acmCert.Arn)
-
-	return acmCert.Arn, nil
-}
 
-func filterExistingARNs(awsAdapter *aws.Adapter, certificateARNs map[string][]*kubernetes.Ingress) (map[string][]*kubernetes.Ingress, map[string]*aws.LoadBalancer) {
-	missingARNs := make(map[string][]*kubernetes.Ingress)
-	existingARNs := make(map[string]*aws.LoadBalancer)
-	for certificateARN, ingresses := range certificateARNs {
-		lb, err := awsAdapter.FindLoadBalancerWithCertificateID(certificateARN)
-		if err != nil && err != aws.ErrLoadBalancerNotFound {
-			log.Println(err)
-			continue
-		}
-		if lb != nil {
-			log.Printf("found existing ALB %q with certificate %q\n", lb.Name(), certificateARN)
-			existingARNs[certificateARN] = lb
-		} else {
-			log.Printf("ALB with certificate %q not found\n", certificateARN)
-			missingARNs[certificateARN] = ingresses
-		}
+	certificateSummary, err := certs.FindBestMatchingCertificate(knownCertificates, ingress.CertHostname())
+	if err != nil {
+		return "", fmt.Errorf("discoverCertificateAndUpdateIngress failed to find a certificate for %q: %v",
+			ingress.CertHostname(), err)
 	}
-	return missingARNs, existingARNs
+	ingress.SetCertificateARN(certificateSummary.ID())
+	return certificateSummary.ID(), nil
 }
 
-func createMissingLoadBalancer(awsAdapter *aws.Adapter, certificateARN string) (*aws.LoadBalancer, error) {
-	log.Printf("creating ALB for ARN %q\n", certificateARN)
+func createStack(awsAdapter *aws.Adapter, item *managedItem) {
+	certificateARN := item.ingress.CertificateARN()
+	log.Printf("creating stack for certificate %q / ingress %q", certificateARN, item.ingress)
 
-	lb, err := awsAdapter.CreateLoadBalancer(certificateARN)
+	stackId, err := awsAdapter.CreateStack(certificateARN)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create ALB for certificate %q. %v\n", certificateARN, err)
+		if isAlreadyExistsError(err) {
+			item.stack, err = awsAdapter.GetStack(stackId)
+			if err == nil {
+				return
```

Comment:
This is implicit: stack = nil, if it's not ready and later it is checked in updateIngress() to return fast (doWork() for-switch statement) and being retriggered in the next loop.